### PR TITLE
Make it possible to use cmpctmalloc for Linux apps for testing

### PR DESCRIPTION
**Pull request by erikcorry**
This has two aims.  Firstly we want to test our embedded software on the desktop, secondly in a later change we can add testing to the malloc implementation, eg causing an arbitrary malloc to return null to see if we handle allocation failure correctly.

### CODE REVIEW

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 1]
```diff
@@ -1340,4 +1347,91 @@ int main(int argc, char *argv[]) {
...
+void dump_on_exit()
+{
+    if (heap != NULL) cmpct_dump(heap);
+}
+
+static cmpct_heap_t *initial_heap()
+{
+    uintptr_t size = 1 << 30;  // 1 Gigabyte.
+    void* pages = NULL;
+    while (pages == NULL) {
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c]: do { } while -- or while (true) + break?

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 2]
```diff
@@ -1340,4 +1347,91 @@ int main(int argc, char *argv[]) {
...
+    if (elsize >= 0x40 && nelem > 0x400) {
+        // Check for multiplication overflow after fast case check failed.
+        const int word_bits = sizeof(size_t) * CHAR_BIT;
+        int bits = word_bits - clzs(nelem);
+        bits +=  word_bits - clzs(elsize);
+        if (bits > word_bits) return NULL;  // Multiplication ran out of bits.
+    }
+    size_t size = nelem * elsize;
+    void *ptr = malloc(size);
+    if (ptr == NULL) return ptr;
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c]: Maybe memset if ptr isn't NULL?

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 3]
```diff
@@ -1340,4 +1347,91 @@ int main(int argc, char *argv[]) {
...
+}
+
+void *calloc(size_t nelem, size_t elsize)
+{
+    if (heap == NULL) heap = initial_heap();
+    if (elsize >= 0x40 && nelem > 0x400) {
+        // Check for multiplication overflow after fast case check failed.
+        const int word_bits = sizeof(size_t) * CHAR_BIT;
+        int bits = word_bits - clzs(nelem);
+        bits +=  word_bits - clzs(elsize);
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c]: Weird spacing after `+=`.

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 4]
```diff
@@ -1340,4 +1347,91 @@ int main(int argc, char *argv[]) {
...
+static int clzs(size_t size)
+{
+    ASSERT(size != 0);  // __builtin_clz is undefined for zero input.
+    if (sizeof(size_t) == sizeof(int)) {
+        return __builtin_clz(size);
+    } else if (sizeof(size_t) == sizeof(long)) {
+        return __builtin_clzl(size);
+    } else if (sizeof(size_t) == sizeof(long long)) {
+        return __builtin_clzll(size);
+    }
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c:1409]: Maybe add a default case that fails?

**Review by kasperl — APPROVED** [review]
Looks good to me.
